### DISCUSSION TITLE
How does rustic implement lock free operation?

### DISCUSSION THREAD

**Discussion by preranaprabhu02** [rustic-rs/rustic#429] [Q&A] [answered]
Can you provide some support material?  I am having a hard time as I am very new to Rust.

**Comment by aawsome** [comment, marked as answer]
Did you read https://github.com/rustic-rs/rustic/blob/main/FAQ.md?
Which additional information do you need?

**Reply by saschmit** [reply]
That URL is now a 404. It can now be located at https://rustic.cli.rs/docs/FAQ.html#how-does-the-lock-free-prune-work

**Comment by preranaprabhu02** [comment]
I had one confusion, in the FAQ about the lock free pruning it states the the packs to be pruned are marked and stored in a separate index structure. Is it stored with the "time" after which it has to be deleted or like how does the time concept work?
Thank you for the FAQ's link, gives good insight

**Reply by aawsome** [reply]
The `prune` run which marks the pack to-be-deleted stores the time at which the pack gets marked (i.e. the current time). 

In the next `prune` run the decision whether to delete that pack is made based on the stored time, the current time and the option `--keep-delete` of that `prune` run.

So, the stored time is not the removal time, but the mark time.

**Reply by preranaprabhu02** [reply]
This makes a sense, but what about the line that stated "_The time to hold marked packs should be long enough to guarantee that a possibly parallel backup run has finished in between. It can be set by the --keep-delete option and defaults to 23 hours. In any case, packs will be kept marked and only deleted by the next prune run."_  so does that mean after 23 hours of the next prune run it deletes these previously marked files?

**Reply by aawsome** [reply]
The marked pack files are removed by the next prune run if
- they are not needed by a backup which finished in between
- the mark time is older than the `--keep-delete` interval which defaults to 23h

Two examples (with `--keep-delete=23h`):

first prune run at 0:00 marks a pack file for removal
second prune run at 14:00 => pack is kept (not old enough), but still marked with the time of the first prune run.
third prune run at 23:30 => pack is removed (if not needed in between)


first prune run at 0:00 marks a pack file for removal
second prune run at 0:00 next day => pack is removed (if not needed in between)

**Reply by preranaprabhu02** [reply]
I understood this properly. Thank you!
